service/header/verify.go

```diff
@@ -0,0 +1,52 @@
+package header
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+)
+
+func VerifyAdjacent(trusted, untrusted *ExtendedHeader) error {
+	if untrusted.Height != trusted.Height+1 {
+		return fmt.Errorf("headers must be adjacent in height")
+	}
+
+	if untrusted.ChainID != trusted.ChainID {
+		return fmt.Errorf("header belongs to another chain %q, not %q", untrusted.ChainID, trusted.ChainID)
+	}
+
+	if untrusted.Height <= trusted.Height {
+		return fmt.Errorf("expected new header height %d to be greater than one of old header %d",
+			untrusted.Height,
+			trusted.Height)
+	}
+
+	if !untrusted.Time.After(trusted.Time) {
+		return fmt.Errorf("expected new header time %v to be after old header time %v",
+			untrusted.Time,
+			trusted.Time)
+	}
+
+	now := time.Now()
+	if !untrusted.Time.Before(now) {
+		return fmt.Errorf("new header has a time from the future %v (now: %v)",
+			untrusted.Time,
+			now)
+	}
+
+	// Check the validator hashes are the same
+	if !bytes.Equal(untrusted.ValidatorsHash, trusted.NextValidatorsHash) {
+		return fmt.Errorf("expected old header next validators (%X) to match those from new header (%X)",
+			trusted.NextValidatorsHash,
+			untrusted.ValidatorsHash,
+		)
+	}
+
+	// Ensure that +2/3 of new validators signed correctly.
+	if err := untrusted.ValidatorSet.VerifyCommitLight(trusted.ChainID, untrusted.Commit.BlockID,
+		untrusted.Height, untrusted.Commit); err != nil {
```

Comment:
nit: I would suggest putting each argument on a new line.

Also: cool, the verify commit logic is re-used from tendermint.

Reply:
Yeah, this part is easy and pleasant to reuse, so why not.